wfe2/wfe.go

```diff
@@ -1053,12 +1053,88 @@ func (wfe *WebFrontEndImpl) setCORSHeaders(response http.ResponseWriter, request
+	outerBody, outerJWS, acct, prob := wfe.validPOSTForAccount(request, ctx, logEvent)
+	addRequesterHeader(response, logEvent.Requester)
+	if prob != nil {
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+
+	// Parse the inner JWS from the validated outer JWS body
+	innerJWS, prob := wfe.parseJWS(outerBody)
+	if prob != nil {
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+
+	// Validate the inner JWS as a key rollover request for the outer JWS
+	rolloverRequest, prob := wfe.validKeyRollover(outerJWS, innerJWS, logEvent)
+	if prob != nil {
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+	newKey := rolloverRequest.NewKey
+
+	// Check that the rollover request's account URL matches the account URL used
+	// to validate the outer JWS
+	accountURL := wfe.jwsKeyIDToAccountID(outerJWS)
+	if rolloverRequest.Account != accountURL {
+		wfe.sendError(response, logEvent, probs.Malformed(
+			fmt.Sprintf("Inner key rollover request specified Account %q, but outer JWS has Key ID %q",
+				rolloverRequest.Account, accountURL)), nil)
+		return
+	}
+
+	// Check that the new key isn't the same as the old key. This would fail as
+	// part of the subsequent `wfe.SA.GetRegistrationByKey` check since the new key
+	// will find the old registration if its equal to the old registration key. We
+	// check new key against old key explicitly to save an RPC round trip and a DB
+	// query for this easy rejection case
+	keysEqual, err := core.PublicKeysEqual(newKey.Key, acct.Key.Key)
+	if err != nil {
+		// This should not happen - both the old and new key have been validated by now
+		wfe.sendError(response, logEvent, probs.ServerInternal("Unable to compare new and old keys"), nil)
+		return
+	}
+	if keysEqual {
+		wfe.sendError(response, logEvent, probs.Malformed(
+			"New key specified by rollover request is the same as the old key"), nil)
+		return
+	}
+
+	// Check that the new key isn't already being used for an existing account
+	if existingAcct, err := wfe.SA.GetRegistrationByKey(ctx, &newKey); err != nil {
+		response.Header().Set("Location", wfe.relativeEndpoint(request, fmt.Sprintf("%s%d", regPath, existingAcct.ID)))
+		wfe.sendError(response, logEvent, probs.Conflict("New key is already in use for a different account"), err)
+		return
+	}
+
+	// Update the account key to the new key
+	updatedAcct, err := wfe.RA.UpdateRegistration(ctx, *acct, core.Registration{Key: &newKey})
+	if err != nil {
+		wfe.sendError(response, logEvent,
+			problemDetailsForError(err, "Unable to update account with new key"), err)
+		return
+	}
+
+	// Marshal & return the new account information
+	jsonReply, err := marshalIndent(updatedAcct)
+	if err != nil {
+		// This should not happen - we've already marshalled the new key by now and
+		// everything else is from the SA
+		wfe.sendError(response, logEvent, probs.ServerInternal("Failed to marshal account"), err)
+		return
+	}
+	response.Header().Set("Content-Type", "application/json")
```

Comment:
Good catch! For whatever reason the original WFE `KeyRollover` [doesn't use `writeJsonResponse`](https://github.com/letsencrypt/boulder/blob/3140c3f16038163682f91a20cae754bbb8508ae4/wfe/wfe.go#L1573:L1581) and I didn't notice when I hacked it up for the WFE2.